Auto Provisioning Kubernetes LoadBalancer Services with F5 (#47)
mesosphere/d2iq-engineering-blog: content/posts/f5-k8s-lb-svc/index.md

6. If using the CAPI steps then a CAPI bootstrap/management cluster and configurations to deploy a cluster. Refer [DKP](https://docs.d2iq.com/dkp/latest/infrastructure-quick-start-guides) documentation for more details.
7. This uses `docker.io/f5networks/f5-ipam-controller:0.1.5` for [FIC](https://github.com/F5Networks/f5-ipam-controller) & `docker.io/f5networks/k8s-bigip-ctlr:2.9.1` for [CIS](https://github.com/F5Networks/k8s-bigip-ctlr) capabilities. Download, retag and push the images to a local registry and change the deployment spec to point to a local image registry for airgapped environments.

This blog used the following versions to test:

* BIG-IP: 16.1.3.1 Build 0.0.11 Point Release 1
* AS3: v3.39.0
* BIG-IP-CTLR: 2.9.1

## Steps to Integrate a Kuberntes cluster with F5 BIG-IP
There are two options based on whether F5 controllers are to be directly deployed to the target cluster, or deployed via [`ClusterResourceSets`](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-resource-set.html) for a [CAPI](https://cluster-api.sigs.k8s.io/) provisioned cluster either at cluster creation time or after the cluster has been deployed.

> <br/>

Reviewer: Do we really need those? Lemme assume we don't and suggest changes...

> <br/>

Reviewer: ```suggestion

```

### Option 1: Directly deploy F5 Controllers to a Kubernetes Cluster

> <br/>

Reviewer: ```suggestion

```

#### Step 1: Deploy F5 Big IP Container Ingress Services (CIS)

<br/>

##### - Add helm repo
```
helm repo add f5-stable https://f5networks.github.io/charts/stable
```

##### - Create values yaml

```
export BIG_IP_URL=https://big-ip-host
export BIG_IP_PARTITION=big-ip-partition
export CLUSTER_NAME=dkp-demo

cat <<EOF > f5-${CLUSTER_NAME}-values.yaml
bigip_login_secret: f5-bigip-ctlr-login
rbac:
  create: true
serviceAccount:
  create: true
namespace: kube-system
args:
  bigip_url: ${BIG_IP_URL}
  bigip_partition: ${BIG_IP_PARTITION}
  log_level: info  
  pool_member_type: nodeport
  insecure: true
  custom-resource-mode: true
  log-as3-response: true
  ipam : true
image:
  # Use the tag to target a specific version of the Controller
  user: f5networks
  repo: k8s-bigip-ctlr
  pullPolicy: Always
resources: {}
version: 2.9.1
EOF
```

##### - Install

```
export F5_USER=f5-user
export F5_PASSWD=f5-password
export KUBECONFIG=kubeconfig-file-path

kubectl create secret generic f5-bigip-ctlr-login -n kube-system --from-literal=username=${F5_USER} --from-literal=password=${F5_PASSWD} 

helm install -f f5-${CLUSTER_NAME}-values.yaml f5ctlr f5-stable/f5-bigip-ctlr --version 0.0.21
```

> <br/>

Reviewer: ```suggestion

```

```

##### - Create values yaml

```
export RANGE='{"ingress":"144.217.53.168-144.217.53.169"}'
cat <<EOF > f5-ipam-${CLUSTER_NAME}-values.yaml
rbac:
  create: true
serviceAccount:
  create: true
namespace: kube-system
args:
  orchestration: "kubernetes"
  provider: "f5-ip-provider"
  ip_range: '$RANGE'
image:
  # Use the tag to target a specific version of the Controller
  user: f5networks
  repo: f5-ipam-controller
  pullPolicy: Always
  version: 0.1.5
volume:
  mountPath: /app/ipamdb
  mountName: fic-volume-mount
  pvc: fic-volume-claim
EOF
```
>The RANGE variable contains key/value pairs for labels and the IP ranges to be served by the IPAM controller. The range used here should be a valid reserved IP range.

##### - Install

```
export KUBECONFIG=kubeconfig-file-path

# Create PVC
kubectl apply -f - <<EOF
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: fic-volume-claim
  namespace: kube-system
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 0.1Gi
EOF

helm install -f f5-ipam-${CLUSTER_NAME}-values.yaml f5-ipam  f5-ipam-stable/f5-ipam-controller --version 0.0.1
```

> <br/>
> <br/>

Reviewer: ```suggestion

```

### Option 2: Deploy Automatically via CAPI 
>Note: If deploying to a CAPI provisioned Kubernetes Cluster like [DKP](https://docs.d2iq.com/dkp/latest/infrastructure-quick-start-guides) instead of running the install command manually, the above can be packaged into a CAPI ClusterResourceSet by doing the following and incorporated into the cluster deployment process.

> <br/>

Reviewer: ```suggestion

```

---
authors: ["ADD AUTHOR NAME"]
title: "Auto Provisioning Kubernetes LoadBalancer Services with F5"
date: 2022-11-17T00:35:56-05:00
featured: false
tags: [loadbalancer]

> excerpt: Learn how to configure a kubernetes cluster to provision a service of type LoadBalancer with F5 BIG-IP 

Reviewer 2: ```suggestion
excerpt: Learn how to configure a Kubernetes cluster to provision a service of type LoadBalancer with F5 BIG-IP 
```

feature_image: feature.png
---

> # Options to provision LoadBalancer Services in an on-premise Kubernetes Cluster

Reviewer 2: ```suggestion
# Options to provision LoadBalancer Services in an on-premise Kubernetes cluster
```

> When deploying Kubernetes clusters in the cloud (AWS, Azure, GCP etc.) a Kubernetes service of type [LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) is by default provisioned automatically by the native load-balancer service (e.g. [ELB](https://aws.amazon.com/elasticloadbalancing/) on AWS). It's not the same for on-premise clusters, which require additional components to be deployed/configured to get this functionality. There are few Kubernetes native options like [metallb](https://metallb.universe.tf/) and [kube-vip](https://kube-vip.io/). However, for production clusters and clusters that process heavy traffic, an external load-balancer might be a better choice. [F5 BIG-IP LTM](https://www.f5.com/products/big-ip-services/local-traffic-manager) is one of the most widely used load-balancers in the industry today and the best thing is that it integrates really nicely with Kubernetes. 

Reviewer 2: ```suggestion
When deploying Kubernetes clusters in the cloud (AWS, Azure, GCP etc.) a Kubernetes service of type [LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) is by default provisioned automatically by the appropriate [cloud controller manager](https://kubernetes.io/docs/concepts/architecture/cloud-controller/) using the native load-balancer service (e.g. [ELB](https://aws.amazon.com/elasticloadbalancing/) on AWS). It's not the same for on-premise clusters, which require additional components to be deployed/configured to get this functionality. There are a few Kubernetes native options like [metallb](https://metallb.universe.tf/) and [kube-vip](https://kube-vip.io/). However, for production clusters and clusters that process heavy traffic, an external load-balancer might be a better choice. [F5 BIG-IP LTM](https://www.f5.com/products/big-ip-services/local-traffic-manager) is one of the most widely used load-balancers in the industry today and the best thing is that it integrates really nicely with Kubernetes. 
```

> In this blog we will see how to configure an on-premise Kubernetes cluster to integrate with [F5 BIG-IP LTM](https://www.f5.com/products/big-ip-services/local-traffic-manager) to provision a service of type LoadBalancer.

Reviewer 2: ```suggestion
In this blog we will see how to configure an on-premise Kubernetes cluster to integrate with [F5 BIG-IP LTM](https://www.f5.com/products/big-ip-services/local-traffic-manager) to provision a service of type `LoadBalancer`.
```

# Integrating a Kubernetes cluster with F5

## Requirements
1. Pre-configured F5 BIG cluster 
2. F5 Partition that will be managed by this automation along with credentials for a service account that has admin permissions for the given partition
3. AS3 3.39 or newer installed on the F5 cluster
4. IP's availble to be used as VIPs
5. Working Kubernetes cluster with PV Storage configured
6. If using the CAPI steps then a CAPI bootstrap/management cluster and configurations to deploy a cluster. Refer [DKP](https://docs.d2iq.com/dkp/latest/infrastructure-quick-start-guides) documentation for more details.

> 7. This uses `docker.io/f5networks/f5-ipam-controller:0.1.5` & `docker.io/f5networks/k8s-bigip-ctlr:2.9.1` images. Download, retag and push the images to a local registry and change the deployment spec to point to a local image registry for airgapped environments.

Reviewer 2: Can you please add links to the different components here? I don't know what they are and not every reader will either so linking would help the reader.

> >Versions used to test:
> >- BIG-IP: 16.1.3.1 Build 0.0.11 Point Release 1
> >- AS3: v3.39.0
> >- BIG-IP-CTLR: 2.9.1

Reviewer 2: ```suggestion
This blog used the following versions to test:

* BIG-IP: 16.1.3.1 Build 0.0.11 Point Release 1
* AS3: v3.39.0
* BIG-IP-CTLR: 2.9.1
```

> ## Steps to Integrate a Kuberntes Cluster with F5 BIG-IP

Reviewer 2: ```suggestion
## Steps to Integrate a Kuberntes cluster with F5 BIG-IP
```

Here are the steps to be performed for a kubernetes cluster that is to be integrated with F5 Big IP to provision services of type LoadBalancer. 

> There are two options based on whether F5 controllers are to be directly deployed to the target cluster or deployed via ClusterResourceSets for a CAPI provisioned cluster either at cluter creation time or after the cluster has been deployed.

Reviewer 2: ```suggestion
There are two options based on whether F5 controllers are to be directly deployed to the target cluster, or deployed via [`ClusterResourceSets`](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-resource-set.html) for a [CAPI](https://cluster-api.sigs.k8s.io/) provisioned cluster either at cluster creation time or after the cluster has been deployed.
```
